<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="ns" default=""/>
  <arg name="rviz" default="true"/>
  <let name="rviz_profile" value="$(find-pkg-share tag_based_sfm_calibrator)/rviz/default.rviz"/>
  <arg name="calibration_service_name" default="extrinsic_calibration"/>

  <arg name="config_file" default="omiya_calibration_room_2023.param.yaml"/>
  <let name="param_path" value="$(find-pkg-share tag_based_sfm_calibrator)/config/$(var config_file)"/>

  <arg name="publish_tfs" default="true"/>
  <arg name="fixed_ground_plane_model" default="false"/>

  <!-- Frame configuration -->
  <arg name="base_frame" default="base_link"/>
  <arg name="main_calibration_sensor_frame"/>

  <!-- Why would a vehicle have more than 8 lidars -->
  <arg name="calibration_lidar_0_frame" default=""/>
  <arg name="calibration_lidar_1_frame" default=""/>
  <arg name="calibration_lidar_2_frame" default=""/>
  <arg name="calibration_lidar_3_frame" default=""/>
  <arg name="calibration_lidar_4_frame" default=""/>
  <arg name="calibration_lidar_5_frame" default=""/>
  <arg name="calibration_lidar_6_frame" default=""/>
  <arg name="calibration_lidar_7_frame" default=""/>

  <arg name="calibration_camera_0_frame" default=""/>
  <arg name="calibration_camera_1_frame" default=""/>
  <arg name="calibration_camera_2_frame" default=""/>
  <arg name="calibration_camera_3_frame" default=""/>
  <arg name="calibration_camera_4_frame" default=""/>
  <arg name="calibration_camera_5_frame" default=""/>
  <arg name="calibration_camera_6_frame" default=""/>
  <arg name="calibration_camera_7_frame" default=""/>

  <let
    name="calibration_lidar_frames"
    value="['$(var calibration_lidar_0_frame)', '$(var calibration_lidar_1_frame)', '$(var calibration_lidar_2_frame)', '$(var calibration_lidar_3_frame)', '$(var calibration_lidar_4_frame)', '$(var calibration_lidar_5_frame)', '$(var calibration_lidar_6_frame)', '$(var calibration_lidar_7_frame)']"
  />
  <let
    name="calibration_camera_frames"
    value="['$(var calibration_camera_0_frame)', '$(var calibration_camera_1_frame)', '$(var calibration_camera_2_frame)', '$(var calibration_camera_3_frame)', '$(var calibration_camera_4_frame)', '$(var calibration_camera_5_frame)', '$(var calibration_camera_6_frame)', '$(var calibration_camera_7_frame)']"
  />

  <!-- Lidar configuration -->
  <arg name="calibration_lidar_0_model" default=""/>
  <arg name="calibration_lidar_1_model" default=""/>
  <arg name="calibration_lidar_2_model" default=""/>
  <arg name="calibration_lidar_3_model" default=""/>
  <arg name="calibration_lidar_4_model" default=""/>
  <arg name="calibration_lidar_5_model" default=""/>
  <arg name="calibration_lidar_6_model" default=""/>
  <arg name="calibration_lidar_7_model" default=""/>

  <!-- Topic configuration -->
  <arg name="calibration_lidar_0_topic" default="/null"/>
  <!--we need a valid topic name due to rename rules -->
  <arg name="calibration_lidar_1_topic" default="/null"/>
  <arg name="calibration_lidar_2_topic" default="/null"/>
  <arg name="calibration_lidar_3_topic" default="/null"/>
  <arg name="calibration_lidar_4_topic" default="/null"/>
  <arg name="calibration_lidar_5_topic" default="/null"/>
  <arg name="calibration_lidar_6_topic" default="/null"/>
  <arg name="calibration_lidar_7_topic" default="/null"/>

  <let name="calibration_lidar_0_detections_topic" value=""/>
  <let name="calibration_lidar_1_detections_topic" value=""/>
  <let name="calibration_lidar_2_detections_topic" value=""/>
  <let name="calibration_lidar_3_detections_topic" value=""/>
  <let name="calibration_lidar_4_detections_topic" value=""/>
  <let name="calibration_lidar_5_detections_topic" value=""/>
  <let name="calibration_lidar_6_detections_topic" value=""/>
  <let name="calibration_lidar_7_detections_topic" value=""/>

  <let name="calibration_lidar_0_detections_topic" value="lidar0/detections_array" if="$(eval &quot;'$(var calibration_lidar_0_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_1_detections_topic" value="lidar1/detections_array" if="$(eval &quot;'$(var calibration_lidar_1_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_2_detections_topic" value="lidar2/detections_array" if="$(eval &quot;'$(var calibration_lidar_2_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_3_detections_topic" value="lidar3/detections_array" if="$(eval &quot;'$(var calibration_lidar_3_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_4_detections_topic" value="lidar4/detections_array" if="$(eval &quot;'$(var calibration_lidar_4_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_5_detections_topic" value="lidar5/detections_array" if="$(eval &quot;'$(var calibration_lidar_5_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_6_detections_topic" value="lidar6/detections_array" if="$(eval &quot;'$(var calibration_lidar_6_frame)' != '' &quot;)"/>
  <let name="calibration_lidar_7_detections_topic" value="lidar7/detections_array" if="$(eval &quot;'$(var calibration_lidar_7_frame)' != '' &quot;)"/>

  <arg name="calibration_camera_info_0_topic" default=""/>
  <arg name="calibration_camera_info_1_topic" default=""/>
  <arg name="calibration_camera_info_2_topic" default=""/>
  <arg name="calibration_camera_info_3_topic" default=""/>
  <arg name="calibration_camera_info_4_topic" default=""/>
  <arg name="calibration_camera_info_5_topic" default=""/>
  <arg name="calibration_camera_info_6_topic" default=""/>
  <arg name="calibration_camera_info_7_topic" default=""/>

  <arg name="calibration_compressed_image_0_topic" default=""/>
  <arg name="calibration_compressed_image_1_topic" default=""/>
  <arg name="calibration_compressed_image_2_topic" default=""/>
  <arg name="calibration_compressed_image_3_topic" default=""/>
  <arg name="calibration_compressed_image_4_topic" default=""/>
  <arg name="calibration_compressed_image_5_topic" default=""/>
  <arg name="calibration_compressed_image_6_topic" default=""/>
  <arg name="calibration_compressed_image_7_topic" default=""/>

  <arg name="calibration_camera_0_detections_topic" default=""/>
  <arg name="calibration_camera_1_detections_topic" default=""/>
  <arg name="calibration_camera_2_detections_topic" default=""/>
  <arg name="calibration_camera_3_detections_topic" default=""/>
  <arg name="calibration_camera_4_detections_topic" default=""/>
  <arg name="calibration_camera_5_detections_topic" default=""/>
  <arg name="calibration_camera_6_detections_topic" default=""/>
  <arg name="calibration_camera_7_detections_topic" default=""/>

  <let name="calibration_camera_0_detections_topic" value="camera0/detections_array" if="$(eval &quot;'$(var calibration_camera_0_frame)' != '' &quot;)"/>
  <let name="calibration_camera_1_detections_topic" value="camera1/detections_array" if="$(eval &quot;'$(var calibration_camera_1_frame)' != '' &quot;)"/>
  <let name="calibration_camera_2_detections_topic" value="camera2/detections_array" if="$(eval &quot;'$(var calibration_camera_2_frame)' != '' &quot;)"/>
  <let name="calibration_camera_3_detections_topic" value="camera3/detections_array" if="$(eval &quot;'$(var calibration_camera_3_frame)' != '' &quot;)"/>
  <let name="calibration_camera_4_detections_topic" value="camera4/detections_array" if="$(eval &quot;'$(var calibration_camera_4_frame)' != '' &quot;)"/>
  <let name="calibration_camera_5_detections_topic" value="camera5/detections_array" if="$(eval &quot;'$(var calibration_camera_5_frame)' != '' &quot;)"/>
  <let name="calibration_camera_6_detections_topic" value="camera6/detections_array" if="$(eval &quot;'$(var calibration_camera_6_frame)' != '' &quot;)"/>
  <let name="calibration_camera_7_detections_topic" value="camera7/detections_array" if="$(eval &quot;'$(var calibration_camera_7_frame)' != '' &quot;)"/>

  <let
    name="calibration_lidar_topics"
    value="['$(var calibration_lidar_0_topic)', '$(var calibration_lidar_1_topic)', '$(var calibration_lidar_2_topic)', '$(var calibration_lidar_3_topic)', '$(var calibration_lidar_4_topic)', '$(var calibration_lidar_5_topic)', '$(var calibration_lidar_6_topic)', '$(var calibration_lidar_7_topic)']"
  />
  <let
    name="calibration_lidar_detections_topics"
    value="['$(var calibration_lidar_0_detections_topic)', '$(var calibration_lidar_1_detections_topic)', '$(var calibration_lidar_2_detections_topic)', '$(var calibration_lidar_3_detections_topic)', '$(var calibration_lidar_4_detections_topic)', '$(var calibration_lidar_5_detections_topic)', '$(var calibration_lidar_6_detections_topic)', '$(var calibration_lidar_7_detections_topic)']"
  />
  <let
    name="calibration_camera_info_topics"
    value="['$(var calibration_camera_info_0_topic)', '$(var calibration_camera_info_1_topic)', '$(var calibration_camera_info_2_topic)', '$(var calibration_camera_info_3_topic)', '$(var calibration_camera_info_4_topic)', '$(var calibration_camera_info_5_topic)', '$(var calibration_camera_info_6_topic)', '$(var calibration_camera_info_7_topic)']"
  />
  <let
    name="calibration_compressed_image_topics"
    value="['$(var calibration_compressed_image_0_topic)', '$(var calibration_compressed_image_1_topic)', '$(var calibration_compressed_image_2_topic)', '$(var calibration_compressed_image_3_topic)', '$(var calibration_compressed_image_4_topic)', '$(var calibration_compressed_image_5_topic)', '$(var calibration_compressed_image_6_topic)', '$(var calibration_compressed_image_7_topic)']"
  />
  <let
    name="calibration_camera_detections_topics"
    value="['$(var calibration_camera_0_detections_topic)', '$(var calibration_camera_1_detections_topic)', '$(var calibration_camera_2_detections_topic)', '$(var calibration_camera_3_detections_topic)', '$(var calibration_camera_4_detections_topic)', '$(var calibration_camera_5_detections_topic)', '$(var calibration_camera_6_detections_topic)', '$(var calibration_camera_7_detections_topic)']"
  />

  <arg
    name="auxiliar_tag_ids"
    default="[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]"
  />
  <arg name="waypoint_tag_ids" default="[0, 3, 4, 5]"/>
  <arg
    name="ground_tag_ids"
    default="[50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]"
  />
  <arg name="left_wheel_tag_id" default="22"/>
  <arg name="right_wheel_tag_id" default="26"/>

  <let name="apriltag_detector_families" value="[16h5, 36h11]"/>

  <group>
    <push-ros-namespace namespace="$(var ns)"/>
    <node pkg="tag_based_sfm_calibrator" exec="tag_based_sfm_calibrator" name="tag_based_sfm_calibrator" output="screen">
      <param from="$(var param_path)"/>
      <remap from="extrinsic_calibration" to="$(var calibration_service_name)"/>

      <param name="publish_tfs" value="$(var publish_tfs)"/>
      <param name="ba.fixed_ground_plane_model" value="$(var fixed_ground_plane_model)"/>

      <!-- Frame configuration -->
      <param name="base_frame" value="$(var base_frame)"/>
      <param name="main_calibration_sensor_frame" value="$(var main_calibration_sensor_frame)"/>
      <param name="calibration_lidar_frames" value="$(var calibration_lidar_frames)"/>
      <param name="calibration_camera_frames" value="$(var calibration_camera_frames)"/>

      <!-- Topic configuration -->
      <param name="calibration_lidar_detections_topics" value="$(var calibration_lidar_detections_topics)"/>
      <param name="calibration_camera_detections_topics" value="$(var calibration_camera_detections_topics)"/>
      <param name="calibration_camera_info_topics" value="$(var calibration_camera_info_topics)"/>
      <param name="calibration_compressed_image_topics" value="$(var calibration_compressed_image_topics)"/>

      <!-- Tag configuration-->
      <param name="waypoint_tag.ids" value="$(var waypoint_tag_ids)"/>
      <param name="ground_tag.ids" value="$(var ground_tag_ids)"/>
      <param name="auxiliar_tag.ids" value="$(var auxiliar_tag_ids)"/>
      <param name="left_wheel_tag_id" value="$(var left_wheel_tag_id)"/>
      <param name="right_wheel_tag_id" value="$(var right_wheel_tag_id)"/>
    </node>

    <node pkg="tag_based_sfm_calibrator" exec="calibrator_ui_node.py" name="calibrator_ui" output="screen"/>

    <node pkg="tf2_ros" exec="static_transform_publisher" name="estimated_base_link_broadcaster" output="screen" args="0 0 0 0 0 0 $(var main_calibration_sensor_frame) estimated_base_link"/>
    <node pkg="tf2_ros" exec="static_transform_publisher" name="main_sensor_broadcaster" output="screen" args="0 0 0 0 0 0 $(var main_calibration_sensor_frame) main_sensor"/>

    <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen" args="-d $(var rviz_profile)">
      <remap from="lidar0_pointcloud" to="$(var calibration_lidar_0_topic)"/>
      <remap from="lidar1_pointcloud" to="$(var calibration_lidar_1_topic)"/>
      <remap from="lidar2_pointcloud" to="$(var calibration_lidar_2_topic)"/>
      <remap from="lidar3_pointcloud" to="$(var calibration_lidar_3_topic)"/>
      <remap from="lidar4_pointcloud" to="$(var calibration_lidar_4_topic)"/>
      <remap from="lidar5_pointcloud" to="$(var calibration_lidar_5_topic)"/>
      <remap from="lidar6_pointcloud" to="$(var calibration_lidar_6_topic)"/>
      <remap from="lidar7_pointcloud" to="$(var calibration_lidar_7_topic)"/>
    </node>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_0_frame)' != '' &quot;)">
      <arg name="ns" value="lidar0"/>
      <arg name="lidar_model" value="$(var calibration_lidar_0_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_0_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_1_frame)' != '' &quot;)">
      <arg name="ns" value="lidar1"/>
      <arg name="lidar_model" value="$(var calibration_lidar_1_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_1_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_2_frame)' != '' &quot;)">
      <arg name="ns" value="lidar2"/>
      <arg name="lidar_model" value="$(var calibration_lidar_2_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_2_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_3_frame)' != '' &quot;)">
      <arg name="ns" value="lidar3"/>
      <arg name="lidar_model" value="$(var calibration_lidar_3_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_3_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_4_frame)' != '' &quot;)">
      <arg name="ns" value="lidar4"/>
      <arg name="lidar_model" value="$(var calibration_lidar_4_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_4_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_5_frame)' != '' &quot;)">
      <arg name="ns" value="lidar5"/>
      <arg name="lidar_model" value="$(var calibration_lidar_5_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_5_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_6_frame)' != '' &quot;)">
      <arg name="ns" value="lidar6"/>
      <arg name="lidar_model" value="$(var calibration_lidar_6_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_6_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/lidartag_detector.launch.xml" if="$(eval &quot;'$(var calibration_lidar_7_frame)' != '' &quot;)">
      <arg name="ns" value="lidar7"/>
      <arg name="lidar_model" value="$(var calibration_lidar_7_model)"/>
      <arg name="pointcloud_topic" value="$(var calibration_lidar_7_topic)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_0_frame)' != '' &quot;)">
      <arg name="ns" value="camera0"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_0_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_0_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_1_frame)' != '' &quot;)">
      <arg name="ns" value="camera1"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_1_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_1_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_2_frame)' != '' &quot;)">
      <arg name="ns" value="camera2"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_2_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_2_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_3_frame)' != '' &quot;)">
      <arg name="ns" value="camera3"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_3_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_3_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_4_frame)' != '' &quot;)">
      <arg name="ns" value="camera4"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_4_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_4_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_5_frame)' != '' &quot;)">
      <arg name="ns" value="camera5"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_5_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_5_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.xml" if="$(eval &quot;'$(var calibration_camera_6_frame)' != '' &quot;)">
      <arg name="ns" value="camera6"/>
      <arg name="image_compressed_topic" value="$(var calibration_compressed_image_6_topic)"/>
      <arg name="camera_info_topic" value="$(var calibration_camera_info_6_topic)"/>
      <arg name="apriltag_detector_families" value="$(var apriltag_detector_families)"/>
      <arg name="filtered_detections_topic" value="detections_array"/>
    </include>
  </group>
</launch>
